accounts-db/src/accounts_file.rs

```diff
@@ -63,49 +66,90 @@ impl AccountsFile {
     /// accounts file.
     pub fn new_from_file(path: impl AsRef<Path>, current_len: usize) -> Result<(Self, usize)> {
-        let (av, num_accounts) = AppendVec::new_from_file(path, current_len)?;
-        Ok((Self::AppendVec(av), num_accounts))
+        match TieredStorage::new_readonly(path.as_ref()) {
+            Ok(tiered_storage) => {
+                // we are doing unwrap here because TieredStorage::new_readonly() is
+                // guaranteed to have a valid reader instance when opening with
+                // new_readonly.
+                let num_accounts = tiered_storage.reader().unwrap().num_accounts();
+                Ok((Self::TieredHot(tiered_storage), num_accounts))
+            }
+            Err(TieredStorageError::MagicNumberMismatch(_, _)) => {
+                // In case of MagicNumberMismatch, we can assume that this is not
+                // a tiered-storage file.
+                let (av, num_accounts) = AppendVec::new_from_file(path, current_len)?;
+                Ok((Self::AppendVec(av), num_accounts))
+            }
+            Err(e) => Err(AccountsFileError::TieredStorageError(e)),
+        }
     }
 
     pub fn flush(&self) -> Result<()> {
         match self {
             Self::AppendVec(av) => av.flush(),
+            Self::TieredHot(_) => Ok(()),
         }
     }
 
     pub fn reset(&self) {
         match self {
             Self::AppendVec(av) => av.reset(),
+            Self::TieredHot(_) => {}
         }
     }
 
     pub fn remaining_bytes(&self) -> u64 {
         match self {
             Self::AppendVec(av) => av.remaining_bytes(),
+            Self::TieredHot(ts) => {
+                if ts.is_read_only() {
+                    0
+                } else {
+                    u64::MAX
+                }
+            }
         }
     }
 
     pub fn len(&self) -> usize {
         match self {
             Self::AppendVec(av) => av.len(),
+            Self::TieredHot(ts) => ts.file_size().unwrap() as usize,
         }
     }
 
     pub fn is_empty(&self) -> bool {
         match self {
             Self::AppendVec(av) => av.is_empty(),
+            Self::TieredHot(ts) => ts.file_size().unwrap() == 0,
         }
     }
 
     pub fn capacity(&self) -> u64 {
         match self {
             Self::AppendVec(av) => av.capacity(),
+            Self::TieredHot(ts) => {
+                // As TieredStorage doesn't have the concept about capacity,
+                // here we return the file size when the file was written.
+                if ts.is_read_only() {
+                    return ts.file_size().unwrap_or(0);
+                }
+                // Or u64::MAX, indicating it can accept as many data as
+                // possible.
+                // TODO: while TieredHot isn't directly capped by size, it
+                // has maximum numbers of accounts and owners.  A proper API
+                // is needed to capture this.
+                u64::MAX
+            }
```

Comment:
It is used to determine how much we can write into a file, and how much we can append.

If the file is already written, then we return the file size here for the tiered-storage.
But if nothing is written yet, then accounts-db needs a number about how much it can store.

Reply:
Here're the direct and indirect call-sites.  It doesn't seem to be used during writes (except tests) -- so good news!

* SplitAncientStore::new()
* AccountsDb::get_unique_accounts_from_storage()
* AccountsDb::select_candidates_by_total_usage()
* AccountsDb::is_shrinking_productive()
* AccountsDb::combine_multiple_slots_into_one_at_startup()
* AccountsDb::purge_dead_slots_from_storage<'a>()